Type Error: Type 'tensor(float16)' of input parameter (/model.23/Constant_23_output_0) of operator (Range) in node (/model.23/Range_4) is invalid

### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.


### Ultralytics YOLO Component

Predict

### Bug

use `yolo11l` onnx to predict get error `[ONNXRuntimeError] : 10 : INVALID_GRAPH : Load model from ../../../data/train/runs/train42/weights/best.onnx failed:This is an invalid model. Type Error: Type 'tensor(float16)' of input parameter (/model.23/Constant_23_output_0) of operator (Range) in node (/model.23/Range_4) is invalid.`

export onnx use
```
def export():
    model = YOLO("runs/train42/weights/best.pt")
    # onnx_file = model.export(format="onnx", dynamic=True, simplify=True, device=0, nms=False)
    onnx_file = model.export(format="onnx", dynamic=True, half=True, device=0,imgsz=1024,simplify=True,opset=20)
    print(onnx_file)
```

log is

```
PyTorch: starting from 'runs/train42/weights/best.pt' with input shape (1, 3, 1024, 1024) BCHW and output shape(s) (1, 90, 21504) (50.8 MB)

ONNX: starting export with onnx 1.17.0 opset 20...
ONNX: slimming with onnxslim 0.1.35...
ONNX: export success ✅ 574.9s, saved as 'runs/train42/weights/best.onnx' (50.2 MB)

Export complete (576.7s)
Results saved to /opt/data/train/runs/train42/weights
Predict:         yolo predict task=obb model=runs/train42/weights/best.onnx imgsz=1024 half 
Validate:        yolo val task=obb model=runs/train42/weights/best.onnx imgsz=1024 data=/opt/datasets/20241218obb/data.yaml half 
Visualize:       https://netron.app
runs/train42/weights/best.onnx
```

### Environment

```
nvidia-cublas-cu12           12.1.3.1
nvidia-cuda-cupti-cu12       12.1.105
nvidia-cuda-nvrtc-cu12       12.1.105
nvidia-cuda-runtime-cu12     12.1.105
nvidia-cudnn-cu12            9.1.0.70
nvidia-cufft-cu12            11.0.2.54
nvidia-curand-cu12           10.3.2.106
nvidia-cusolver-cu12         11.4.5.107
nvidia-cusparse-cu12         12.1.0.106
nvidia-nccl-cu12             2.20.5
nvidia-nvjitlink-cu12        12.1.105
nvidia-nvtx-cu12             12.1.105
onnx                         1.17.0
onnx-graphsurgeon            0.5.2
onnx2tf                      1.22.3
onnxruntime-gpu              1.20.1
onnxslim                     0.1.35
ultralytics                  8.3.15       /ultralytic
```

### Minimal Reproducible Example

```
conf = 0.6
obb = True
agnostic_nms = True
iou = 0.3
device = 0


def show_result(frame, window_name="Detection Results"):
    cv2.namedWindow(window_name, cv2.WINDOW_FULLSCREEN)
    cv2.imshow(window_name, frame)
    key = cv2.waitKey(0) & 0xFF
    return key


def video_infer(filepath, weight_path):
    try:
        model = YOLO(weight_path,task='detect')
        imgsz = 1024
        if os.path.isdir(filepath):
            for file in tqdm(os.listdir(filepath)):
                if file.endswith('.jpg') or file.endswith('.PNG'):
                    path = os.path.join(filepath, file)
                    results = model(path,
                                    conf=conf,
                                    iou=iou,
                                    imgsz=imgsz,
                                    half=True,
                                    device=device,
                                    max_det=50,
                                    show=False,
                                    save=False,
                                    show_labels=True,
                                    show_conf=True,
                                    agnostic_nms=agnostic_nms,
                                    )
                    if obb:
                        boxes = results[0].obb
```

### Additional

_No response_

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

It works for me on the latest version of Ultralytics. Also the latest version shows this error when using both half and dynamic:

```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[16], line 1
----> 1 model.export(format="onnx", dynamic=True, half=True)

File ~/ultralytics/ultralytics/engine/model.py:738, in Model.export(self, **kwargs)
    730 custom = {
    731     "imgsz": self.model.args["imgsz"],
    732     "batch": 1,
   (...)
    735     "verbose": False,
    736 }  # method defaults
    737 args = {**self.overrides, **custom, **kwargs, "mode": "export"}  # highest priority args on the right
--> 738 return Exporter(overrides=args, _callbacks=self.callbacks)(model=self.model)

File ~/ultralytics/ultralytics/engine/exporter.py:266, in Exporter.__call__(self, model)
    264     LOGGER.warning("WARNING ⚠️ half=True only compatible with GPU export, i.e. use device=0")
    265     self.args.half = False
--> 266     assert not self.args.dynamic, "half=True not compatible with dynamic=True, i.e. use only one."
    267 self.imgsz = check_imgsz(self.args.imgsz, stride=model.stride, min_dim=2)  # check image size
    268 if self.args.int8 and engine:

AssertionError: half=True not compatible with dynamic=True, i.e. use only one.
```

--- a/docs/en/reference/engine/exporter.md
+++ b/docs/en/reference/engine/exporter.md
@@ -37,4 +37,8 @@ keywords: YOLOv8, export formats, ONNX, TensorRT, CoreML, machine learning model
 
 ## ::: ultralytics.engine.exporter.try_export
 
+<br><br><hr><br>
+
+## ::: ultralytics.engine.exporter.arange_patch
+
 <br><br>

--- a/ultralytics/__init__.py
+++ b/ultralytics/__init__.py
@@ -1,6 +1,6 @@
 # Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license
 
-__version__ = "8.3.81"
+__version__ = "8.3.82"
 
 import os
 

--- a/ultralytics/engine/exporter.py
+++ b/ultralytics/engine/exporter.py
@@ -62,6 +62,7 @@ import shutil
 import subprocess
 import time
 import warnings
+from contextlib import contextmanager
 from copy import deepcopy
 from datetime import datetime
 from pathlib import Path
@@ -183,7 +184,7 @@ def try_export(inner_func):
     return outer_func
 
 
-@__import__("contextlib").contextmanager
+@contextmanager
 def arange_patch(args):
     """
     Workaround for ONNX torch.arange incompatibility with FP16.

--- a/ultralytics/engine/exporter.py
+++ b/ultralytics/engine/exporter.py
@@ -186,7 +186,7 @@ def try_export(inner_func):
 @__import__("contextlib").contextmanager
 def arange_patch(args):
     """Workaround for ONNX torch.arange incompatibility with FP16."""
-    if args.dynamic and args.half:
+    if args.dynamic and args.half and args.format == "onnx":
         func = torch.arange
 
         def arange(*args, dtype=None, **kwargs):

--- a/ultralytics/engine/exporter.py
+++ b/ultralytics/engine/exporter.py
@@ -185,7 +185,7 @@ def try_export(inner_func):
 
 @__import__("contextlib").contextmanager
 def arange_patch(args):
-    """Workaround for ONNX torch.arange incompatibility with FP16."""
+    """Workaround for ONNX torch.arange incompatibility with FP16. https://github.com/pytorch/pytorch/issues/148041"""
     if args.dynamic and args.half and args.format == "onnx":
         func = torch.arange
 

--- a/ultralytics/engine/exporter.py
+++ b/ultralytics/engine/exporter.py
@@ -185,7 +185,7 @@ def try_export(inner_func):
 
 @__import__("contextlib").contextmanager
 def arange_patch(args):
-    """Workaround for ONNX torch.arange incompatibility with FP16. https://github.com/pytorch/pytorch/issues/148041"""
+    """Workaround for ONNX torch.arange incompatibility with FP16. https://github.com/pytorch/pytorch/issues/148041."""
     if args.dynamic and args.half and args.format == "onnx":
         func = torch.arange
 

--- a/ultralytics/engine/exporter.py
+++ b/ultralytics/engine/exporter.py
@@ -183,6 +183,22 @@ def try_export(inner_func):
     return outer_func
 
 
+@__import__("contextlib").contextmanager
+def arange_patch(args):
+    """Workaround for ONNX torch.arange incompatibility with FP16."""
+    if args.dynamic and args.half:
+        func = torch.arange
+
+        def arange(*args, dtype=None, **kwargs):
+            return func(*args, **kwargs).to(dtype)  # workaround arange FP16 incompatibility
+
+        torch.arange = arange  # patch
+        yield
+        torch.arange = func  # unpatch
+    else:
+        yield
+
+
 class Exporter:
     """
     A class for exporting a model.
@@ -263,7 +279,6 @@ class Exporter:
         if self.args.half and onnx and self.device.type == "cpu":
             LOGGER.warning("WARNING ⚠️ half=True only compatible with GPU export, i.e. use device=0")
             self.args.half = False
-            assert not self.args.dynamic, "half=True not compatible with dynamic=True, i.e. use only one."
         self.imgsz = check_imgsz(self.args.imgsz, stride=model.stride, min_dim=2)  # check image size
         if self.args.int8 and engine:
             self.args.dynamic = True  # enforce dynamic to export TensorRT INT8
@@ -515,7 +530,6 @@ class Exporter:
         output_names = ["output0", "output1"] if isinstance(self.model, SegmentationModel) else ["output0"]
         dynamic = self.args.dynamic
         if dynamic:
-            self.model.cpu()  # dynamic=True only compatible with cpu
             dynamic = {"images": {0: "batch", 2: "height", 3: "width"}}  # shape(1,3,640,640)
             if isinstance(self.model, SegmentationModel):
                 dynamic["output0"] = {0: "batch", 2: "anchors"}  # shape(1, 116, 8400)
@@ -533,17 +547,18 @@ class Exporter:
                 pass
             check_requirements("onnxslim>=0.1.46")  # Older versions has bug with OBB
 
-        torch.onnx.export(
-            NMSModel(self.model, self.args) if self.args.nms else self.model,
-            self.im.cpu() if dynamic else self.im,
-            f,
-            verbose=False,
-            opset_version=opset_version,
-            do_constant_folding=True,  # WARNING: DNN inference with torch>=1.12 may require do_constant_folding=False
-            input_names=["images"],
-            output_names=output_names,
-            dynamic_axes=dynamic or None,
-        )
+        with arange_patch(self.args):
+            torch.onnx.export(
+                NMSModel(self.model, self.args) if self.args.nms else self.model,
+                self.im,
+                f,
+                verbose=False,
+                opset_version=opset_version,
+                do_constant_folding=True,  # WARNING: DNN inference with torch>=1.12 may require do_constant_folding=False
+                input_names=["images"],
+                output_names=output_names,
+                dynamic_axes=dynamic or None,
+            )
 
         # Checks
         model_onnx = onnx.load(f)  # load onnx model

--- a/ultralytics/engine/exporter.py
+++ b/ultralytics/engine/exporter.py
@@ -185,7 +185,11 @@ def try_export(inner_func):
 
 @__import__("contextlib").contextmanager
 def arange_patch(args):
-    """Workaround for ONNX torch.arange incompatibility with FP16. https://github.com/pytorch/pytorch/issues/148041."""
+    """
+    Workaround for ONNX torch.arange incompatibility with FP16.
+
+    https://github.com/pytorch/pytorch/issues/148041.
+    """
     if args.dynamic and args.half and args.format == "onnx":
         func = torch.arange
 

--- a/ultralytics/engine/exporter.py
+++ b/ultralytics/engine/exporter.py
@@ -190,7 +190,7 @@ def arange_patch(args):
         func = torch.arange
 
         def arange(*args, dtype=None, **kwargs):
-            return func(*args, **kwargs).to(dtype)  # workaround arange FP16 incompatibility
+            return func(*args, **kwargs).to(dtype)  # cast to dtype instead of passing dtype
 
         torch.arange = arange  # patch
         yield

